### DISCUSSION TITLE
List Article Mdx

### DISCUSSION THREAD

**Discussion by enri90** [vikejs/vike#1082] [Discussion] [answered]
how can i create the article list of the mdx pages?

https://vite-plugin-ssr.com/markdown#react

i created the various pages using the guide now i have to show the list of those pages with one pagination

https://blog.logrocket.com/create-next-js-mdx-blog/
in other guides they use to read the post folder with its mdx and extract the meta data such as publication date, author, and category

Thank you

**Comment by brillout** [comment]
Added some docs about this: ebd84766a4c45e0d66fb1fa431924de52ea45645 (rendered: https://vite-plugin-ssr.com/markdown#page-list). (CC @jrson83.)

Let me know if you still have questions.

**Comment by FlorianCassayre** [comment]
An alternative to the `metadata.ts` file that I have found to work well is to organize your articles following this structure:
- `my-article-slug/`
  - `+Page.tsx`: the article content
  - `+config.ts`: the article metadata (title, etc.)

The `+config.ts` files are then automatically aggregated using a glob import. With this approach, lazy-transpilation still occurs because the article metadata is decoupled from its content.

The advantages that I find over the recommended approach are:
- Colocation: it's easy to make the connection between the article metadata and its content
- DRY: this eliminates the need of a `url` attribute, as it is implicitly deduced from the file location
- Efficiency: the article can access its own metadata more efficiently, no need to load the other articles' metadata

**Reply by brillout** [reply]
Good point and, indeed, it's something that is now possible with the V1 design (it wasn't possible before).

> using a glob import

This shouldn't be necessary: I'm thinking of officially exposing `globalContext.pages` that includes everything Vike knows about pages. That will do the trick and no need to use a glob import then. (It's actually good timing as I'm currently working on an offical data structure to expose these kind of information.)

See also: https://vike.dev/meta.

Let's eventually update the docs https://github.com/vikejs/vike/issues/2113 — after `globalContext.pages` is shipped (possibly already soon).

Thank you!

**Reply by FlorianCassayre** [reply]
> > using a glob import
>
> This shouldn't be necessary

Good observation, I didn't think of that and it might in fact be slightly cleaner.

I can spend some time on a draft PR for #2113 later in the week.

**Reply by brillout** [reply]
That would be great indeed and I was thinking of labeling it as `contribution-welcome`. But let me first finish what I've in mind with `globalContext.pages`. I'll keep you updated. (Contributions about other things are very much welcome btw.)

**Reply by brillout** [reply]
> But let me first finish what I've in mind with `globalContext.pages`

ETA this week. (It's mostly done, I just need to expose it.)

**Reply by brillout** [reply]
> `globalContext.pages`

It's now available with the latest version. Either over `pageContext.globalContext.pages` or using `getGlobalContextAsync()`.

Example of using it: https://github.com/vikejs/vike/blob/d808eda1a262e01785d3413e59a8e691909ff21e/test/playground/pages/%2BonBeforeRender.ts#L5-L13

But for the idea to fully come to fruition we'd need https://github.com/vikejs/vike/issues/962. I'll try to work on it sooner rather than latter.

**Reply by FlorianCassayre** [reply]
Perfect, thanks, I'll check this out when I have time.

**Reply by brillout** [reply]
Done.

https://vike.dev/globalContext
https://vike.dev/globalContext#pages
https://vike.dev/onCreateGlobalContext

That provides all primitives you will/may need to achieve what we discussed.

I'll be updating the guides accordingly (once I'm done with my current priorities, or PR much welcome).

The playground I linked to in my previous comment is updated: it's now accurate and the recommended way.

**Reply by brillout** [reply, marked as answer]
Updated docs: https://vike.dev/markdown#with-a-custom-setting-eager.

**Reply by teddybradford** [reply]
> Updated docs: https://vike.dev/markdown#with-a-custom-setting-eager.

@brillout Can this work using [frontmatter](https://vike.dev/markdown#with-frontmatter) instead of metadata files?

**Reply by brillout** [reply]
@teddybradford Not sure what you mean — these two approaches aren't complementary, it's either or.

**Reply by teddybradford** [reply]
If, for example, I wanted to render a list of posts using the frontmatter data inside all my `.md` files, what's the ideal way to do that with Vike?

**Reply by brillout** [reply]
See warning at https://vike.dev/markdown#global-metadata

**Reply by teddybradford** [reply]
Ohh, I see. I misunderstood the warning when I first read it. Maybe it could be more explicit like:

> You may be tempted to use [import.meta.glob()](https://vitejs.dev/guide/features.html#glob-import) to retrieve _frontmatter metadata from_ all pages but …

**Reply by brillout** [reply]
:+1: Should be clearer now: https://github.com/vikejs/vike/commit/12603dc0799ff70306ff6cbfc8360a1f37215e46
